### PR TITLE
WIP: Import policies in a policy

### PR DESCRIPTION
**Pull request by vhdirk**
This is an implementation for #298 (referenced by #991)

### CODE REVIEW

**Comment by thjaeckle** [comment]
@vhdirk I will try to merge "master" onto my 2 year old experiment of adding policyImports - that should give a really good start for putting additional required work on top of it (e.g. how to handle search index updates then).

**Comment by vhdirk** [comment]
Oh, that'd be very cool! Thanks!

**Comment by thjaeckle** [comment]
I pushed a branch: https://github.com/eclipse/ditto/tree/lab/policy-imports
I invested a little trying to merge master onto that - I think from Policy Model side I got very far - however there still is a lot of stuff not even compiling - I did not even come to the tricky part the "policy enforcement" and the policy cache loading / cache invalidation, etc.

To be honest: I think without deep understanding of the Ditto codebase this will be really hard to get finished.

**Comment by thjaeckle** [comment]
@vhdirk I got the build working - did some first rough code adjustments right on your branch, I hope you don't mind.
Will look into the functionality tomorrow.

What still is missing are unit tests, e.g. for:
* ImmutableEffectedImports
* ImmutableImportedLabels
* ImmutablePolicyImport
* ImmutablePolicyImports
* LabelInvalidException
* PolicyImportHelper
And also for:
* PolicyCacheLoader

**Comment by thjaeckle** [comment]
@vhdirk I added a "feature toggle" for the policy import feature - we recently already added something like this for the "merge things" feature.

I started Ditto on this feature branch, however the imported policies were not yet applied - I will now look into this, but I fear that currently the policy import is not yet working.

**Comment by thjaeckle** [comment]
@vhdirk I invested some time in getting the functionality to actually "work".
* For direct access to a thing/policy (knowing their IDs) I would say that the policy import does work perfectly

We have *a few open topics before being able to merge this PR*.

# search index

However, the search index is only correctly updated for the "happy path":
* when a `ThingUpdater` responsible for updating a thing with ID `x:y-1` is currently "alive" (is running), updating a policy which is imported in other policies will correctly update the things using the imported policy
   * so this is only the case when changes were recently done on a thing and e.g. not on a cluster restart

This is how the search index is updated when a policy is changed:
1. the `PolicyEvent` is received in `PolicyEventForwarder` of the search service
2. `PolicyEventForwarder` finds out all things which (directly) use this policy (via `ThingsSearchUpdaterPersistence.getPolicyReferenceTags(Map)` and creates for each matched thing one `PolicyReferenceTag` message
3. each created `PolicyReferenceTag` message is locally forwarded to `ThingsUpdater`
4. the `ThingsUpdater` invalidates the `policyCache` of that policy (on the local "search" instance)
5. the `ThingsUpdater` sends each `PolicyReferenceTag` via cluster sharding to a `ThingUpdater` in a search instance "responsible" for the shard of the thingId (consistently always the same instance for the same ID)

The problem is now that step `2.` does not yet respect imported policies when calculating the `PolicyReferenceTag` message based on the update of a policy which was imported.
* therefore, we would need to adjust the MongoDB collection `searchThings` (or introduce another persistent MongoDB collection) in which such information is persisted
* basically we must answer the question: "which things are affected (because they link to a policy 'B' importing policy 'A') when policy 'A' is modified?"
* in the implementation `MongoThingsSearchUpdaterPersistence.getPolicyReferenceTags(Map)` those affected policyIds must additionally be returned as `PolicyReferenceTag`s

When enhancing the search index to know that we:
* must ensure to stay backwards compatible (existing indexed things must still be found the same as before)
* must look into MongoDB indexes - so that those queries are fast
* have the freedom of add something like `additionalPolicyIds` - as the search index is "only" used in order to determine thingIds matching a search query .. the structure of it is never showed to the Ditto API user

# importable

And one more thing we found in a peer-review with the team:
* a (potentially to be imported) policy must be able to declare single entries as "not importable" - otherwise when importing a complete policy there could be entries imported which are not desired to "include"
* as when creating existing policies this had not to be thought about we think that it's "safer" to assume that existing policies shall not be imported

We propose to add an optional `"importable": true` (the default value to assume if the field is not present is `false` - following the *principle of least privilege*) JSON field to the Policy Entry, e.g.:
```json
{
  "policyId": "foo:bar",
  "entries": {
    "SUPER_ADMIN": {
      "importable": false,
      "subjects": {
        ...
      },
      "resources": {
        ...
       },
...
```

# summing up
1. Search must be fixed to consistently update things in the index when imported policies are changed
2. Support *principle of least privilege* by adding "importable" enum to PolicyEntry defaulting to `false` - that way importing must be explicitly enabled
3. Documentation (OpenAPI + Ditto website docs) is missing
4. Additional (integration) tests testing applying the policies, e.g. added to `ThingCommandEnforcementTest`

I would say we are 70% through :-)

File: services/thingsearch/persistence/src/main/java/org/eclipse/ditto/services/thingsearch/persistence/write/streaming/EnforcementFlow.java [hunk 1]
```diff
@@ -298,4 +315,76 @@ private static boolean shouldReloadCache(@Nullable final Entry<?> entry, final M
...
+                        return POLICY_NONEXISTENT;
+                    });
+
+            return Source.completionStageSource(policyFuture);
+        });
+
+        return lazySource.viaMat(Flow.create(), Keep.none());
+    }
+
+    private Source<List<Entry<Policy>>, NotUsed> readCachedImportedPolicies(final Metadata metadata,
```
Comment (vhdirk) [inline services/thingsearch/persistence/src/main/java/org/eclipse/ditto/services/thingsearch/persistence/write/streaming/EnforcementFlow.java:367]: @thjaeckle I think this is somewhat like you meant on gitter?

Comment (thjaeckle) [reply services/thingsearch/persistence/src/main/java/org/eclipse/ditto/services/thingsearch/persistence/write/streaming/EnforcementFlow.java:384]: @vhdirk hm, almost :)
I don't get why you load the complete policy of the "imported policies" - for the `List<PolicyTag> policyImports`  passed into the `EnforcedThingMapper.toWriteModel(..)` method you don't need to load the `Policy`, the id of the imported policies + the revision number is sufficient.
I would simplify the return types of the private "read..." methods. Also, I don't think that you have to use `Entry` there as wrapping cache entry type - just use the `PolicyTag` or the list of dertermined `PolicyTag`s.

I would expect a method signature:
```
private Source<List<PolicyTag>, NotUsed> getPolicyImports(final PolicyId policyId);
```
Which can simply be called in addition to `getEnforcer` to get input for the to be called `EnforcedThingMapper.toWriteModel()`.

Comment (vhdirk) [reply services/thingsearch/persistence/src/main/java/org/eclipse/ditto/services/thingsearch/persistence/write/streaming/EnforcementFlow.java:384]: Ok, I'll refactor it. You weren't kidding when you said that this requires messing about with some deep internals :p

Comment (vhdirk) [reply services/thingsearch/persistence/src/main/java/org/eclipse/ditto/services/thingsearch/persistence/write/streaming/EnforcementFlow.java:384]: I suppose it also doesn't quite make sense to reload the cached entry? https://github.com/vhdirk/ditto/blob/7d57d4f590300b14d242cdb299d31beaf04c90db/services/thingsearch/persistence/src/main/java/org/eclipse/ditto/services/thingsearch/persistence/write/streaming/EnforcementFlow.java#L362

Comment (vhdirk) [reply services/thingsearch/persistence/src/main/java/org/eclipse/ditto/services/thingsearch/persistence/write/streaming/EnforcementFlow.java:384]: While refactoring, I think I do need to load at least the thing policy; it may be that a cached version of a policy that is imported by the 'current' policy already exists. We won't know that unless we fetch the latest revision of the 'current' policy, or do we?

Comment (thjaeckle) [reply services/thingsearch/persistence/src/main/java/org/eclipse/ditto/services/thingsearch/persistence/write/streaming/EnforcementFlow.java:384]: Good point 👍 
The `Metadata` object contains the revision number of the used policy of a Thing update processed in `ThingUpdater`.
However at this point it can't be known whether any of the referenced imported policies have also changed. Even if all cache invalidator callbacks would work as expected, due to the sharding when horizontally scaling the "things-search" service, this information is not consistently known everywhere.

So I think you're right: when a policy uses imported policies, the imported policies (to be precise: their revision number) must be loaded OR the cache entries in the `policyCache` for those must be invalidated which will cause an automatic new loading of them.

This would make it a very expensive (bandwith and CPU consumption wise for deserialization) operation via the "normal" `policyCache` - as this cache always asks for the full Policy JSON.

TL;DR: So I fear we need another cache which holds `PolicyTag`s.

Comment (vhdirk) [reply services/thingsearch/persistence/src/main/java/org/eclipse/ditto/services/thingsearch/persistence/write/streaming/EnforcementFlow.java]: And for such a new cache to work, I guess we're also going to need a new `sudo` command, `SudoRetrievePolicyRevision` or so? Or, perhaps `SudoRetrievePolicyRevisions`, since we know it's going to be a list.

Comment (thjaeckle) [reply services/thingsearch/persistence/src/main/java/org/eclipse/ditto/services/thingsearch/persistence/write/streaming/EnforcementFlow.java]: Yes. You mean like this already existing one `org.eclipse.ditto.services.models.policies.commands.sudo.SudoRetrievePolicyRevision`? ;)

Comment (vhdirk) [reply services/thingsearch/persistence/src/main/java/org/eclipse/ditto/services/thingsearch/persistence/write/streaming/EnforcementFlow.java]: And to know I did actually look for something like that. Time for more coffee :)
Aside from that; this MR is getting pretty big. Do you think we could/need split it up somehow?

Comment (thjaeckle) [reply services/thingsearch/persistence/src/main/java/org/eclipse/ditto/services/thingsearch/persistence/write/streaming/EnforcementFlow.java]: I don't think that splitting up is useful - we wouldn't want to merge a half-working feature to our master branch.
Documentation and integration tests could be provided in another PR afterwards.

And with Ditto 2.0 to be released next week, there will be some (huge, I fear) merge efforts as well on the overall PR - I can support on that.

**Comment by thjaeckle** [comment]
@vhdirk @BobClaerhout I rebased this PR on the current Ditto master branch (took quite some effort because of some breaking Ditto 2.0 module changes) - I referenced @vhdirk as co-author in the commit.

Summarizing there are still some open TODOs:

* My comment https://github.com/eclipse/ditto/pull/993#discussion_r615890894 still is to be addressed - however I think that it would be easier to achieve when simply invalidating imported and affected policies cluster-wide when a policy was changed which was used in other policies as import
   * see my "// TODO TJ" comment in `ThingsUpdater.processPolicyReferenceTag()`
* Documentation (OpenAPI + Ditto website docs) is missing
* Additional (integration) tests testing applying the policies, e.g. added to `ThingCommandEnforcementTest`

**Comment by thjaeckle** [comment]
Superseded by #1555
